core/trino-spi/src/main/java/io/trino/spi/predicate/Range.java

```diff
@@ -27,178 +39,316 @@
+        return getLow().compareTo(marker) <= 0 && getHigh().compareTo(marker) >= 0;
     }
 
     public boolean contains(Range other)
     {
         checkTypeCompatibility(other);
-        return this.getLow().compareTo(other.getLow()) <= 0 &&
-                this.getHigh().compareTo(other.getHigh()) >= 0;
+
+        return compareLowBound(other) <= 0 &&
+                compareHighBound(other) >= 0;
     }
 
     public Range span(Range other)
     {
         checkTypeCompatibility(other);
-        Marker lowMarker = Marker.min(low, other.getLow());
-        Marker highMarker = Marker.max(high, other.getHigh());
-        return new Range(lowMarker, highMarker);
+        int compareLowBound = compareLowBound(other);
+        int compareHighBound = compareHighBound(other);
+        return new Range(
+                type,
+                compareLowBound <= 0 ? this.lowInclusive : other.lowInclusive,
+                compareLowBound <= 0 ? this.lowValue : other.lowValue,
+                compareHighBound >= 0 ? this.highInclusive : other.highInclusive,
+                compareHighBound >= 0 ? this.highValue : other.highValue);
     }
 
-    public boolean overlaps(Range other)
+    public Range intersect(Range other)
     {
         checkTypeCompatibility(other);
-        return this.getLow().compareTo(other.getHigh()) <= 0 &&
-                other.getLow().compareTo(this.getHigh()) <= 0;
+        if (!this.overlaps(other)) {
+            throw new IllegalArgumentException("Cannot intersect non-overlapping ranges");
+        }
+
+        int compareLowBound = compareLowBound(other);
+        int compareHighBound = compareHighBound(other);
+        return new Range(
+                type,
+                compareLowBound >= 0 ? this.lowInclusive : other.lowInclusive,
+                compareLowBound >= 0 ? this.lowValue : other.lowValue,
+                compareHighBound <= 0 ? this.highInclusive : other.highInclusive,
+                compareHighBound <= 0 ? this.highValue : other.highValue);
     }
 
-    public Range intersect(Range other)
+    public boolean overlaps(Range other)
     {
         checkTypeCompatibility(other);
-        if (!this.overlaps(other)) {
-            throw new IllegalArgumentException("Cannot intersect non-overlapping ranges");
+        return !this.isFullyBefore(other) && !other.isFullyBefore(this);
+    }
+
+    /**
+     * Returns unioned range if {@code this} and {@code next} overlap or are adjacent.
+     * The {@code next} lower bound must not be before {@code this} lower bound.
+     */
+    Optional<Range> tryOverlapWithNext(Range next)
+    {
+        if (this.compareLowBound(next) > 0) {
+            throw new IllegalArgumentException("next before this");
+        }
+
+        if (this.isHighUnbounded()) {
+            return Optional.of(this);
+        }
+
+        if (next.isLowUnbounded()) {
+            return Optional.of(next);
```

Comment:
so here you assume that next.high >= this.high? Check?

Reply:
Good catch, this is actually a bug in `RangeView` where this is copied from. I am adding a prep commit.

Reply:
Fix also extracted into https://github.com/trinodb/trino/pull/6895 for visibility